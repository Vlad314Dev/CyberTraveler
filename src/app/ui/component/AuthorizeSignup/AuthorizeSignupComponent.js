import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

class AuthorizeSignupComponent extends PureComponent
{
    static propTypes = {
        handleSubmit: PropTypes.func.isRequired,
        handleNickname: PropTypes.func.isRequired,
        handlePassword: PropTypes.func.isRequired,
        handlePasswordRepeat: PropTypes.func.isRequired
    };

    render()
    {
        const { 
            handleSubmit,
            handleNickname,
            handlePassword,
            handlePasswordRepeat
        } = this.props;

        return (
            <div block="AuthorizeSignup">
                <form onSubmit={ handleSubmit } block="Form">
                    <div block="Form" elem="Field">
                        <label htmlFor="nickname">Nickname:</label>
                        <input type="text" name="nickname" id="nickname" placeholder="Nickname" onChange={ handleNickname } />
                    </div>
                    <div block="Form" elem="Field">
                        <label htmlFor="password">Password:</label>
                        <input type="password" name="password" id="password" onChange={ handlePassword } />
                    </div>
                    <div block="Form" elem="Field">
                        <label htmlFor="password_repeat">Repeat Password:</label>
                        <input type="password" name="password_repeat" id="password_repeat" onChange={ handlePasswordRepeat } />
                    </div>
                    <div block="Form" elem="Submit">
                        <button type="submit">Signup</button>
                    </div>
                </form>
            </div>
        );
    }
}

export default AuthorizeSignupComponent;
