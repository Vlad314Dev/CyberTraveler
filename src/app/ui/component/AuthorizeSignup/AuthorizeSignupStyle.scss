.AuthorizeSignup {
    &-Field {
        text-align: right;
        padding: 0 10px 10px;

        label {
            color: #fff;
            padding-right: 10px;
        }
    }

    &-Submit {
        text-align: center;

        button {
            color: #000;
            padding: 12px 28px;
            text-align: center;
            text-decoration: none;
            font-size: 16px;
            margin: 4px 2px;
            transition-duration: .4s;
            cursor: pointer;
            background-color: #ffc107;

            &:hover {
                background-color: #4caf50;
                color: #fff;
            }
        }
    }
}
