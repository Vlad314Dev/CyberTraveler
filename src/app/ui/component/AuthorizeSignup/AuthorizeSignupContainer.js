// import { PropTypes } from 'prop-types';
import { sha256 } from 'crypto-hash';
import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import GqlClient, {
    SIGNUP
} from 'UIQuery';

import AuthorizeSignupComponent from './AuthorizeSignupComponent';

class AuthorizeSignupContainer extends PureComponent
{
    constructor(props)
    {
        super(props);
        
        this.state = {
            nickname: '',
            password: '',
            passwordRepeat: '',
            loading: false,
            isValidated: false
        }
    }

    containerFunctions = {
        handleSubmit: this.handleSubmit.bind(this),
        handleNickname: this.handleNickname.bind(this),
        handlePassword: this.handlePassword.bind(this),
        handlePasswordRepeat: this.handlePasswordRepeat.bind(this)
    };
    
    handleNickname(e)
    {
        this.setState({
            nickname: e.target.value
        });
    }

    handlePassword(e)
    {
        this.setState({
            password: e.target.value
        });
    }

    handlePasswordRepeat(e)
    {
        this.setState({
            passwordRepeat: e.target.value
        });
    }

    validateFormData()
    {
        const {
            nickname,
            password,
            passwordRepeat
        } = this.state;

        if (!nickname || !password) {
            alert('Error.');
            return false;
        }

        if (password !== passwordRepeat) {
            alert('Passwords do not match.');
            return false;
        }

        if (nickname.length < 3 || nickname.length > 15) {
            alert('Nickname has to be 3-15 characters long.');
            return false;
        }

        if (password.length < 6) {
            alert('Password is too short.');
            return false;
        }

        if (password.length > 255) {
            alert('Password is too long.');
            return false;
        }

        return true;
    }

    async handleSubmit(e)
    {        
        e.preventDefault();
        if (this.validateFormData()) {
            this.setState({
                loading: true
            });
    
            try {
                await this.signUp();
                alert('Success.');
            } catch (err) {
                alert('Something went wrong. Please try again.');
            }
    
            this.setState({
                loading: false
            });
        }
    }

    async signUp()
    {
        const {
            nickname,
            password
        } = this.state;
        const hashedPassword = await sha256(password);
        
        await GqlClient.mutate({
            mutation: SIGNUP,
            variables: {
                nickname: nickname,
                password: hashedPassword
            },
            fetchPolicy: 'no-cache'
        }).then(
            ({ error, data }) => {
                this.setState({
                    error,
                    data
                });
            }
        );
    }

    render()
    {
        return (
            <AuthorizeSignupComponent 
                { ...this.props }
                { ...this.state }
                { ...this.containerFunctions }
            />
        );
    }
}

export default connect(null, null)(AuthorizeSignupContainer);
