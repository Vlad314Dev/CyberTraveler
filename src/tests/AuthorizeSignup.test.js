/* eslint-disable no-undef */
import { shallow } from 'enzyme';
import React from 'react';
import configureMockStore from 'redux-mock-store';

import AuthorizeSignup from '../app/ui/component/AuthorizeSignup';
import AuthorizeSignupComponent from '../app/ui/component/AuthorizeSignup/AuthorizeSignupComponent';

const mockStore = configureMockStore();
const store = mockStore({});

describe('AuthorizeSignup', () => {
    it('Should render', () => {
        // eslint-disable-next-line no-unused-vars
        const component = shallow(
            <AuthorizeSignup 
                store={ store }
            />
        );
    });

    it('Should render layout', () => {
        const component = shallow(
            <AuthorizeSignup 
                store={ store }
            />
        ).dive().find('AuthorizeSignupComponent').dive();

        expect(component.getElements()).toMatchSnapshot();
    });

    it('Should update component state on nickname change', () => {
        const container = shallow(
            <AuthorizeSignup 
                store={ store }
            />
        ).dive();
        const component = container.find('AuthorizeSignupComponent').dive();
        const input = component.find('#nickname');
        const testValue = 'test';
        
        input.props().onChange({
            target: {
                value: testValue
            }
        });

        expect(container.state('nickname')).toEqual(testValue);
    });

    it('Should update component state on password change', () => {
        const container = shallow(
            <AuthorizeSignup 
                store={ store }
            />
        ).dive();
        const component = container.find('AuthorizeSignupComponent').dive();
        const input = component.find('#password');
        const testValue = 'test';
        
        input.props().onChange({
            target: {
                value: testValue
            }
        });

        expect(container.state('password')).toEqual(testValue);
    });

    it('Should update component state on password repeat change', () => {
        const container = shallow(
            <AuthorizeSignup 
                store={ store }
            />
        ).dive();
        const component = container.find('AuthorizeSignupComponent').dive();
        const input = component.find('#password_repeat');
        const testValue = 'test';
        
        input.props().onChange({
            target: {
                value: testValue
            }
        });

        expect(container.state('passwordRepeat')).toEqual(testValue);
    });

    it('Should submit form', () => {
        const testProps = {
            handleSubmit: jest.fn(),
            handleNickname: jest.fn(),
            handlePassword: jest.fn(),
            handlePasswordRepeat: jest.fn(),
            loading: false
        };
        
        const component = shallow(
            <AuthorizeSignupComponent
                store={ store }
                { ...testProps }
            />
        );
        const nicknameInput = component.find('#nickname');
        const passwordInput = component.find('#password');
        const form = component.find('form');

        nicknameInput.props().onChange({
            target: {
                value: 'test'
            }
        });
        passwordInput.props().onChange({
            target: {
                value: 'test'
            }
        });
        
        form.simulate('submit');

        expect(testProps.handleSubmit).toHaveBeenCalled();
    });

    it('Should show alert on form validation error', async () => {
        const container = shallow(
            <AuthorizeSignup 
                store={ store }
            />
        ).dive();
        const component = container.find('AuthorizeSignupComponent').dive();
        const nicknameInput = component.find('#nickname');
        const passwordInput = component.find('#password');
        const passwordRepeatInput = component.find('#password_repeat');
        const form = component.find('form');

        nicknameInput.props().onChange({
            target: {
                value: 'test'
            }
        });
        passwordInput.props().onChange({
            target: {
                value: 'test'
            }
        });
        passwordRepeatInput.props().onChange({
            target: {
                value: 'test'
            }
        });

        window.alert = jest.fn();
        await form.props().onSubmit({ preventDefault: jest.fn() });

        expect(window.alert).toHaveBeenCalled();
    });
});
