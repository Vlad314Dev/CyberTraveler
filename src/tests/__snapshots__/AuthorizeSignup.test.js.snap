// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorizeSignup Should render layout 1`] = `
Array [
  <Unknown
    color="#fff"
    component="div"
    loaded={true}
    loadedClassName="loadedContent"
    parentClassName="loader"
  >
    <div
      className="AuthorizeSignup"
    >
      <form
        className="Form"
        onSubmit={[Function]}
      >
        <div
          className="Form__Field"
        >
          <label
            htmlFor="nickname"
          >
            Nickname:
          </label>
          <input
            id="nickname"
            name="nickname"
            onChange={[Function]}
            placeholder="Nickname"
            type="text"
          />
        </div>
        <div
          className="Form__Field"
        >
          <label
            htmlFor="password"
          >
            Password:
          </label>
          <input
            id="password"
            name="password"
            onChange={[Function]}
            type="password"
          />
        </div>
        <div
          className="Form__Field"
        >
          <label
            htmlFor="password_repeat"
          >
            Repeat Password:
          </label>
          <input
            id="password_repeat"
            name="password_repeat"
            onChange={[Function]}
            type="password"
          />
        </div>
        <div
          className="Form__Submit"
        >
          <button
            type="submit"
          >
            Signup
          </button>
        </div>
      </form>
    </div>
  </Unknown>,
]
`;
